### PR TITLE
Sia20 robot model ikfast

### PR DESCRIPTION
**Pull request by jrgnicho**
The goal of this PR is to use the motoman sia20d implementation of descartes::RobotModel  alongside the descarte planners in order to generate smoother path plans significantly faster than with the current approach, which uses the more generic MoveitStateAdapter implementation.  The details of this PR are as follows:
- Added  moveit::RobotModel plugin for the motoman sia20d  in the motoman_sia20d_descartes package which uses ikfast for that robot
- Added the ikfast pluging for the sia20d as well
- Created a unit test to load the sia20d descartes robot model as a plugin.

To build the unit test run the following in the catkin workspace:

`catkin_make motoman_sia20d_descartes_utest
`

To execute the unit test run the following:

`catkin_make run_tests_motoman_sia20d_descartes_rostest
`

Note: This PR uses the latest version of descartes hydro-devel

### CODE REVIEW

File: motoman_sia20d_descartes/include/motoman_sia20d_descartes/motoman_sia20d_robot_model.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+#ifndef MOTOMAN_SIA20D_ROBOT_MODEL
+#define MOTOMAN_SIA20D_ROBOT_MODEL
+
+#include <descartes_moveit/moveit_state_adapter.h>
+#include <motoman_sia20d_ikfast_manipulator/motoman_sia20d_manipulator_ikfast_moveit_plugin.hpp>
+
+namespace motoman_sia20d_descartes
+{
+  class MotomanSia20dRobotModel: public descartes_moveit::MoveitStateAdapter,
```
Comment (shaun-edwards) [inline motoman_sia20d_descartes/include/motoman_sia20d_descartes/motoman_sia20d_robot_model.h:25]: It feels like this needs to be a base class or maybe a template.  The code in these methods is always the same for each ikfast solution.

File: motoman_sia20d_descartes/test/motoman_sia20d_descartes_plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+static const std::string WORLD_FRAME = "world_frame";
+static const std::string TCP_FRAME = "tool0";
+static const std::string ROBOT_DESCRIPTION = "robot_description";
+Eigen::Affine3d tcp_pose;
+
+pluginlib::ClassLoader<descartes_core::RobotModel> robot_model_loader("descartes_core",
+                                                                     "descartes_core::RobotModel");
+descartes_core::RobotModelPtr robot_model_ptr;
+
+TEST(MotomanSia20dRobotModel, createPluginInstance)
```
Comment (shaun-edwards) [inline motoman_sia20d_descartes/test/motoman_sia20d_descartes_plugin.cpp:17]: This test could be generic for all ikfast solutions (see the other test fixtures).

**Comment by shaun-edwards** [comment]
This is the same bad solution.  Is there enough time/resources to solve this the "right" way.

**Comment by gavanderhoorn** [comment]
Could the idea behind this PR be detailed a bit more perhaps?

I don't understand why the state adapter is also an IKFast plugin. Is there something that prevents the use of the IKFast plugin as a plugin for the state adapter (ie: that is technically impossible because of how `pluginlib` works or something)?

**Comment by shaun-edwards** [comment]
> Is there something that prevents the use of the IKFast plugin as a plugin for the state adapter (ie: that is technically impossible because of how pluginlib works or something)?

The plugin is missing a `getAllIK` method.  See ros-industrial-consortium/descartes#53 and relevant [discussion](https://groups.google.com/forum/#!msg/moveit-users/IGkUhU2CGu0/WADuA85GSmgJ).

**Comment by gavanderhoorn** [comment]
~~Just me being dense here (and you've most likely already seen this), but:~~

``` cpp
/** Computes all IK solutions given a end effector coordinates and the free joints.
 [..]
 */
IKFAST_API bool ComputeIk(const IkReal* eetrans, const IkReal* eerot, const IkReal* pfree, ikfast::IkSolutionListBase<IkReal>& solutions)
```

~~would be nice to use I guess, but the symbol seems not be exported in any of the MoveIt IKFast plugins.~~

~~Is it not an option to compile in the generated `X_manipulator_ikfast_solver.cpp`? Access through the normal MoveIt plugin interface would be nicer.~~

~~Edit: seems `IKFastKinematicsPlugin::solve(..)` actually calls that method.~~

Edit: ignore. Me being dense.

**Comment by jrgnicho** [comment]
@shaun-edwards  I'm guessing that by "the right way" you mean extending the KinematicsBase interface to expose a 'getAllIk' method.  If this is so then time would be the biggest constrain in the context of this project.  But I do agree that this solution is hackish at best.

File: motoman_sia20d_descartes/test/motoman_sia20d_descartes_plugin.cpp (hunk 2 above)
Comment (jrgnicho) [inline motoman_sia20d_descartes/test/motoman_sia20d_descartes_plugin.cpp:17]: Yes, that would help standardize the robot model test

**Comment by jrgnicho** [comment]
@gavanderhoorn did we clarify the root of the issue for you?

**Comment by shaun-edwards** [comment]
@jrgnicho, what you suggest is the "most right way".  The "less right way" would be to minimize the amount of copy/pasting.  For example, this file: `motoman_sia20d_descartes/src/motoman_sia20d_robot_model.cpp` is nearly always a copy and paste with just the robot name/namespace changed.

**Comment by jrgnicho** [comment]
@shaun-edwards , perhaps a template base approach as you suggested would make calls to the ikfast functions in the same way for each robot specific implementation.  Since the automatically generated ikfast code is rather difficult to work with then coming up with a proper generalized procedure/template may not be so straight forward but certainly doable. 

However, since we are currently working towards a "most right solution" I'm not so sure it'd be worth spending time making the fix described above.

**Comment by phvass** [comment]
I talked to Dave Coleman about recently added IKFast capability, which returns the IK that is closest to the current configuration. He mentioned that there were some bugs to fix for that. Paul

Sent from my Verizon Wireless 4G LTE DROID
On Feb 22, 2015 2:19 PM, Jorge Nicho notifications@github.com wrote:

@shaun-edwardshttps://github.com/shaun-edwards , perhaps a template base approach as you suggested would make calls to the ikfast functions in the same way for each robot specific implementation. Since the automatically generated ikfast code is rather difficult to work with then coming up with a proper generalized procedure/template may not be so straight forward but certainly doable.

However, since we are currently working towards a "most right solution" I'm not so sure it'd be worth spending time making the fix described above.

##

**Comment by jrgnicho** [comment]
@Jmeyer1292  see @phvass  comment above when you test this branch.  Thanks

**Pull request merged**